### PR TITLE
VZ-5681: Create top level VMO install component

### PR DESCRIPTION
**Pull request by bobdonat**
# Description

This pull request creates a top level VMO install component so it can be installed separately from the verrazzano install component. 

# Checklist 

As the author of this PR, I have:

- [ ] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/helm_config/charts/verrazzano/values.yaml [hunk 1]
```diff
@@ -89,19 +89,6 @@ api:
   name: verrazzano-authproxy
   port: 8775
 
-# OCI-related values
```
Comment (pfmackin) [inline platform-operator/helm_config/charts/verrazzano/values.yaml:92]: why did you remove OCI settings?  I don't see VMO settings being removed.

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go [hunk 2]
```diff
@@ -151,46 +149,28 @@ func reassociateResources(cli clipkg.Client) error {
 
 	// namespaced resources
 	for _, obj := range objects {
-		if _, err := removeResourcePolicyAnnotation(cli, obj, namespacedName); err != nil {
+		if _, err := common.RemoveResourcePolicyAnnotation(cli, obj, namespacedName); err != nil {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:152]: Why are we touching authproxy resources?

File: platform-operator/controllers/verrazzano/component/authproxy/authproxy.go [hunk 3]
```diff
@@ -151,46 +149,28 @@ func reassociateResources(cli clipkg.Client) error {
...
 			return err
 		}
 	}
 
 	// additional namespaced resources managed by this helm chart
 	authProxyResources := GetHelmManagedResources()
 	for _, managedResoure := range authProxyResources {
-		if _, err := removeResourcePolicyAnnotation(cli, managedResoure.Obj, managedResoure.NamespacedName); err != nil {
+		if _, err := common.RemoveResourcePolicyAnnotation(cli, managedResoure.Obj, managedResoure.NamespacedName); err != nil {
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/authproxy/authproxy.go:160]: Why are we touching authproxy resources?

File: platform-operator/controllers/verrazzano/component/common/helm.go [hunk 4]
```diff
@@ -13,3 +16,50 @@ type HelmManagedResource struct {
...
+		if errors.IsNotFound(err) {
+			return obj, nil
+		}
+		return obj, err
+	}
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		return obj, nil
+	}
+	delete(annotations, "helm.sh/resource-policy")
```
Comment (pfmackin) [inline platform-operator/controllers/verrazzano/component/common/helm.go:61]: Don't you need to delete these?
	annotations["meta.helm.sh/release-name"] = releaseName.Name
	annotations["meta.helm.sh/release-namespace"] = releaseName.Namespace

File: platform-operator/verrazzano-bom.json [hunk 5]
```diff
@@ -185,6 +185,39 @@
...
+          ]
+        }
+      ]
+    },
+    {
+      "name": "verrazzano-monitoring-operator",
+      "subcomponents": [
+        {
+          "repository": "verrazzano",
+          "name": "verrazzano-monitoring-operator",
```
Comment (pfmackin) [inline platform-operator/verrazzano-bom.json:219]: I think the only images this should have is the VMO image and the Proxy image.  The other images should stay in Verrazzano and will soon be moved to other components, like Opensearch.

File: platform-operator/helm_config/charts/verrazzano/values.yaml (hunk 1 above)
Comment (bobdonat) [reply platform-operator/helm_config/charts/verrazzano/values.yaml:92]: Because these settings are not used.  This stuff has been there since the beginning and has nothing to do with the OCI Logging integration.

File: platform-operator/controllers/verrazzano/component/common/helm.go (hunk 4 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/common/helm.go:61]: No, those settings are needed.  Helm installed resources need these annotations.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go [hunk 6]
```diff
@@ -130,6 +123,7 @@ func isVerrazzanoReady(ctx spi.ComponentContext) bool {
 		if ctx.EffectiveCR().Spec.Components.Elasticsearch != nil {
 			esInstallArgs := ctx.EffectiveCR().Spec.Components.Elasticsearch.ESInstallArgs
 			for _, args := range esInstallArgs {
+				/* TODO: this needs be moved to the OS component IsReady check along with the other VMI checks
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: Was this moved?

File: platform-operator/verrazzano-bom.json (hunk 5 above)
Comment (bobdonat) [reply platform-operator/verrazzano-bom.json:219]: VMO needs all these other images.  They are referenced in the VMO helm templates.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_test.go [hunk 7]
```diff
@@ -211,30 +212,6 @@ func TestFixupFluentdDaemonset(t *testing.T) {
 	a.Nil(updatedDaemonSet.Spec.Template.Spec.Containers[0].Env[1].ValueFrom)
 }
 
-// Test_appendCustomImageOverrides tests the appendCustomImageOverrides function
```
Comment (kminder) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_test.go:214]: Why was this custom image override stuff removed?

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go (hunk 6 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: The check was failing because an additional es-data deployment is added during upgrade but it had not been created when the vz component checks are made.  This should be moved during Ben's changes and the check will work.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_test.go (hunk 7 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_test.go:214]: It is now part of the VMO component overrides since these values are needed by the VMO helm chart templates

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go (hunk 6 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: Looks like Ben's preliminary PR has moved this check to the OS component

File: platform-operator/helm_config/charts/verrazzano-monitoring-operator/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+              value: {{ .Values.monitoringOperator.esImage }}
+            - name: ELASTICSEARCH_INIT_IMAGE
+              value: {{ .Values.monitoringOperator.esInitImage }}
+            - name: KIBANA_IMAGE
+              value: {{ .Values.monitoringOperator.kibanaImage }}
+            - name: ELASTICSEARCH_WAIT_TARGET_VERSION
+              value: {{ .Values.monitoringOperator.esWaitTargetVersion }}
+            - name: CONFIG_RELOADER_IMAGE
+              value: {{ .Values.monitoringOperator.configReloaderImage }}
+            - name: OIDC_PROXY_IMAGE
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-monitoring-operator/templates/deployment.yaml:63]: Are all of these still required?

File: platform-operator/helm_config/charts/verrazzano-monitoring-operator/values.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  alertManagerImage: "noimage"
+  esWaitTargetVersion: 1.2.3
+  oidcAuthEnabled: true
+  RequestMemory: 48Mi
+
+config:
+  envName:
+  dnsSuffix:
+
+api:
```
Comment (kminder) [inline platform-operator/helm_config/charts/verrazzano-monitoring-operator/values.yaml:24]: Is this still needed?

Comment (bobdonat) [reply platform-operator/helm_config/charts/verrazzano-monitoring-operator/values.yaml:24]: Yes, the VMO charts reference them.

File: platform-operator/helm_config/charts/verrazzano-monitoring-operator/templates/deployment.yaml (hunk 8 above)
Comment (bobdonat) [reply platform-operator/helm_config/charts/verrazzano-monitoring-operator/templates/deployment.yaml:63]: I don't think so.  I did not make any change to the VMO templates.  I can take another pass and clean this.

File: platform-operator/helm_config/charts/verrazzano/values.yaml (hunk 1 above)
Comment (pfmackin) [reply platform-operator/helm_config/charts/verrazzano/values.yaml:92]: I still see VMO stuff in values. yaml

**Review by pfmackin — DISMISSED** [review]

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go [hunk 10]
```diff
@@ -96,6 +97,12 @@ func (c verrazzanoComponent) Install(ctx spi.ComponentContext) error {
 
 // PreUpgrade Verrazzano component pre-upgrade processing
 func (c verrazzanoComponent) PreUpgrade(ctx spi.ComponentContext) error {
+	if err := vmo.ExportVmoHelmChart(ctx); err != nil {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:100]: So, component ordering during upgrade ensures that this runs first right?  So if we ever move towards a parallel impl this might break correct?  That being said, not sure if there's a better way to handle it now.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go (hunk 6 above)
Comment (kminder) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: got it

Comment (kminder) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: Should the comment be removed then?

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_test.go (hunk 7 above)
Comment (kminder) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_test.go:214]: got it

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go [hunk 11]
```diff
@@ -96,6 +97,12 @@ func (c verrazzanoComponent) Install(ctx spi.ComponentContext) error {
...
+		return err
+	}
+	if err := common.ApplyCRDYaml(ctx, config.GetHelmVmoChartsDir()); err != nil {
```
Comment (michael-cico) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:103]: We have apply this here instead of in the VMO component because it won't run during upgrade right?  Do we have to version-check this call, for post-1.3 when the VMO comp will run during upgrade?  That's assuming we want to have the VMO comp do this work post-1.3

File: platform-operator/helm_config/charts/verrazzano-monitoring-operator/templates/deployment.yaml (hunk 8 above)
Comment (kminder) [reply platform-operator/helm_config/charts/verrazzano-monitoring-operator/templates/deployment.yaml:63]: Another pass in a different PR is probably a good idea since we are touching these.

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go (hunk 10 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:100]: yes and yes

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go (hunk 6 above)
Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: Yes this is moved

Comment (mbenmclean) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano.go:126]: Yes this is moved

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_values.go [hunk 12]
```diff
@@ -27,13 +27,11 @@ type verrazzanoValues struct {
 	Kiali                  *kialiValues                  `json:"kiali,omitempty"`
 	Keycloak               *keycloakValues               `json:"keycloak,omitempty"`
 	Rancher                *rancherValues                `json:"rancher,omitempty"`
-	MonitoringOperator     *vmoValues                    `json:"monitoringOperator,omitempty"`
 	NodeExporter           *nodeExporterValues           `json:"nodeExporter,omitempty"`
 	Logging                *loggingValues                `json:"logging,omitempty"`
 	Fluentd                *fluentdValues                `json:"fluentd,omitempty"`
 	Console                *consoleValues                `json:"console,omitempty"`
 	API                    *apiValues                    `json:"api,omitempty"`
-	OCI                    *ociValues                    `json:"oci,omitempty"`
```
Comment (mbenmclean) [inline platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_values.go:36]: Why is this removed?

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go (hunk 11 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:103]: VMO was not being upgraded and this needed to be run before we updated VMI which is in the verrazzano component.  Not sure whether we need to version-check this call. For post-1.3 we need to call this in pre-upgrade.

File: platform-operator/controllers/verrazzano/component/vmo/vmo.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+			Name:      ComponentName,
+			Namespace: ComponentNamespace,
+		},
+	}
+	prefix := fmt.Sprintf("Component %s", context.GetComponent())
+	return status.DeploymentsAreReady(context.Log(), context.Client(), deployments, 1, prefix)
+}
+
+// appendVmoOverrides appends overrides for the VMO component
+func appendVmoOverrides(ctx spi.ComponentContext, _ string, _ string, _ string, kvs []bom.KeyValue) ([]bom.KeyValue, error) {
```
Comment (mbenmclean) [inline platform-operator/controllers/verrazzano/component/vmo/vmo.go:32]: Should be appendVMOOverrides

File: platform-operator/controllers/verrazzano/component/vmo/vmo.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+	}
+
+	kvs = append(kvs, imageOverrides...)
+	return kvs, nil
+}
+
+// ExportVmoHelmChart adds necessary annotations to verrazzano-monitoring-operator objects which allows them to be
+// managed by the verrazzano-monitoring-operator helm chart.  This is needed for the case when VMO was
+// previously installed by the verrazzano helm charrt.
+func ExportVmoHelmChart(ctx spi.ComponentContext) error {
```
Comment (mbenmclean) [inline platform-operator/controllers/verrazzano/component/vmo/vmo.go:75]: Should be ExportVMOHelmChart

File: platform-operator/controllers/verrazzano/component/vmo/vmo_component.go [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+
+// Verify that vmoComponent implements Component
+var _ spi.Component = vmoComponent{}
+
+// NewComponent returns a new VMO component
+func NewComponent() spi.Component {
+	return vmoComponent{
+		helm.HelmComponent{
+			ReleaseName:             ComponentName,
+			ChartDir:                filepath.Join(config.GetHelmChartsDir(), ComponentName),
```
Comment (mbenmclean) [inline platform-operator/controllers/verrazzano/component/vmo/vmo_component.go:39]: You should set the JSONName field

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_values.go (hunk 12 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_values.go:36]: these values were obsolete and not used

File: platform-operator/controllers/verrazzano/component/vmo/vmo_component.go (hunk 15 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/vmo/vmo_component.go:39]: I did not set it because there is no VMO component in the CRD.  Do I still need it?

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go (hunk 11 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:103]: I can version check the call and add  the call in VMO pre-upgrade. I would like to do that in a follow up PR since others are depending on this merge.

File: platform-operator/controllers/verrazzano/component/vmo/vmo.go (hunk 13 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/vmo/vmo.go:32]: I can change in followup PR

File: platform-operator/controllers/verrazzano/component/vmo/vmo.go (hunk 14 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/vmo/vmo.go:75]: I can change in followup PR

File: platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go (hunk 11 above)
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/verrazzano/verrazzano_component.go:103]: yup, no worries

File: platform-operator/controllers/verrazzano/component/vmo/vmo_component.go (hunk 15 above)
Comment (bobdonat) [reply platform-operator/controllers/verrazzano/component/vmo/vmo_component.go:42]: fixed

File: platform-operator/helm_config/charts/verrazzano/values.yaml (hunk 1 above)
Comment (bobdonat) [reply platform-operator/helm_config/charts/verrazzano/values.yaml:92]: Will address in followup PR

**Review by mbenmclean — APPROVED** [review]
